conda-store-server/conda_store_server/server/views/api.py

```diff
@@ -688,6 +688,60 @@ async def api_put_build(
     }
 
 
+@router_api.put(
+    "/build/{build_id}/cancel/",
+    response_model=schema.APIAckResponse,
+)
+async def api_put_build_cancel(
+    build_id: int,
+    request: Request,
+    conda_store=Depends(dependencies.get_conda_store),
+    db: Session = Depends(dependencies.get_db),
+    auth=Depends(dependencies.get_auth),
+):
+    build = api.get_build(db, build_id)
+    if build is None:
+        raise HTTPException(status_code=404, detail="build id does not exist")
+
+    auth.authorize_request(
+        request,
+        f"{build.environment.namespace.name}/{build.environment.name}",
+        {Permissions.BUILD_CANCEL},
+        require=True,
+    )
+
+    if conda_store.celery_app.control.inspect().ping() is None:
+        raise HTTPException(
+            status_code=409,
+            detail="conda-store celery broker does not support task cancelation. Use redis or rabbitmq message queues. See docs for a more detailed explanation",
+        )
+
+    conda_store.celery_app.control.revoke(
+        [
+            f"build-{build_id}-conda-env-export",
+            f"build-{build_id}-conda-pack",
+            f"build-{build_id}-conda-docker",
+            f"build-{build_id}-environment",
+        ],
+        terminate=True,
+        signal="SIGTERM",
+    )
+
+    from conda_store_server.worker import tasks
+
+    tasks.task_cleanup_builds.si(
+        build_ids=[build_id],
+        reason=f"""
+build {build_id} marked as FAILED due to being canceled from the REST API
+""",
+    ).apply_async(countdown=5)
+
+    return {
+        "status": "ok",
+        "message": f"build {build_id} canceled",
+    }
```

Comment:
Agree. There needs to be a clear status distinction between builds that were cancelled by the user and failed ones.

Reply:
I agree with this and checking celery and redis queues they both have a concept of a canceled task (revoked and canceled).

This will require a lot more file/changes. Since CANCELED is adding a new final state. Also will require an alembic migration to account for this new state.

Should we wrap this into this PR?

Reply:
I was confused by the above thread a bit since the message already says "canceled". The requested change here is to pass a yet-to-be-added CANCELED (made up name) marker/param to `task_cleanup_builds`, to make use of it in `build_cleanup`, which currently just calls `set_build_failed` regardless of the reason.

Will be done in a separate PR.